common/src/main/java/com/wynntils/core/chat/transcoder/CodedStyle.java

```diff
@@ -0,0 +1,175 @@
+/*
+ * Copyright © Wynntils 2023.
+ * This file is released under AGPLv3. See LICENSE for full license details.
+ */
+package com.wynntils.core.chat.transcoder;
+
+import com.wynntils.utils.colors.CustomColor;
+import java.util.Arrays;
+import java.util.Optional;
+import net.minecraft.ChatFormatting;
+import net.minecraft.network.chat.ClickEvent;
+import net.minecraft.network.chat.HoverEvent;
+import net.minecraft.network.chat.Style;
+
+public final class CodedStyle {
+    private static final String STYLE_PREFIX = "§";
+
+    private final CodedStringPart owner;
+
+    private final CustomColor color;
+    private final boolean bold;
+    private final boolean italic;
+    private final boolean underlined;
+    private final boolean strikethrough;
+    private final boolean obfuscated;
+    private final ClickEvent clickEvent;
+    private final HoverEvent hoverEvent;
+
+    private Style styleCache;
+
+    private CodedStyle(
+            CodedStringPart owner,
+            CustomColor color,
+            boolean bold,
+            boolean italic,
+            boolean underlined,
+            boolean strikethrough,
+            boolean obfuscated,
+            ClickEvent clickEvent,
+            HoverEvent hoverEvent) {
+        this.owner = owner;
+        this.color = color;
+        this.bold = bold;
+        this.italic = italic;
+        this.underlined = underlined;
+        this.strikethrough = strikethrough;
+        this.obfuscated = obfuscated;
+        this.clickEvent = clickEvent;
+        this.hoverEvent = hoverEvent;
+    }
+
+    static CodedStyle fromStyle(Style style, CodedStringPart owner) {
+        return new CodedStyle(
+                owner,
+                style.getColor() == null
+                        ? CustomColor.NONE
+                        : CustomColor.fromInt(style.getColor().getValue()),
+                style.isBold(),
+                style.isItalic(),
+                style.isUnderlined(),
+                style.isStrikethrough(),
+                style.isObfuscated(),
+                style.getClickEvent(),
+                style.getHoverEvent());
+    }
+
+    public String asString() {
+        // Rules of converting a Style to a String:
+        // 0. Every style is prefixed with a §.
+        // 1. Style color is converted to a color segment.
+        //    A color segment is the prefix and the chatFormatting char.
+        //    If this is a custom color, a hex color code is used.
+        //    Example: §#FF0000 or §1
+        // 2. Formatting is converted the same way as in the Style class.
+        // 3. Click events are wrapped in angle brackets, and is reprenseted as an id.
+        //    The parent of this style's owner is responsible for keeping track of click events.
+        //    Example: §<1> -> (1st click event)
```

Comment:
Just for the sake of my sanity, can we swap click and hover? I'm thinking of `[1]` like a URL,e.g. like in markdown, so I think of that as an onclick event, rather than a hover.

Reply:
Sure.